File: crates/docpie/src/auth/services.rs
use crate::{auth::models::User, error::Result, http::config::Server};
use password_auth::generate_hash;
use time::OffsetDateTime;

pub async fn get_user_by_id(server: &Server, user_id: i32) -> Result<Option<User>> {
    let user = sqlx::query_as!(
        User,
        r#"
        SELECT
            id,
            email,
            password,
            display_name,
            created_at,
            is_onboarded
        FROM users
        WHERE id = $1
        "#,
        user_id
    )
    .fetch_optional(&server.db)
    .await?;

    Ok(user)
}

pub async fn get_user_by_email(server: &Server, email: String) -> Result<Option<User>> {
    let user = sqlx::query_as!(
        User,
        r#"
        SELECT
            id,
            email,
            password,
            display_name,
            created_at,
            is_onboarded
        FROM users
        WHERE email = $1
        "#,
        email
    )
    .fetch_optional(&server.db)
    .await?;

    Ok(user)
}

pub async fn create_user(
    server: &Server,
    email: &String,
    password: &String,
    display_name: &String,
) -> Result<User> {
    let password = password.to_owned();
    let password = server
        .handle
        .spawn_blocking(move || generate_hash(password))
        .await?;

    let user = sqlx::query_as!(
        User,
        r#"
        INSERT INTO users (email, password, display_name, created_at)
        VALUES ($1, $2, $3, $4)
        RETURNING id, email, password, display_name, created_at, is_onboarded
        "#,
        email,
        password,
        display_name,
        OffsetDateTime::now_utc()
    )
    .fetch_one(&server.db)
    .await?;

    Ok(user)
}

pub async fn onboard_user(server: &Server, user: User) -> Result<User> {
    let user = sqlx::query_as!(
        User,
        r#"
        UPDATE users
        SET is_onboarded = true
        WHERE id = $1
        RETURNING id, email, password, display_name, created_at, is_onboarded
        "#,
        user.id
    )
    .fetch_one(&server.db)
    .await?;

    Ok(user)
}

File: shared/auth/services.ts
import { db } from "#shared/db/client"
import { users, type TUser, type TNewUser } from "#shared/db/schema"
import { hashPassword, verifyPassword } from "#shared/utils/password"
import { eq } from "drizzle-orm"

/**
 * Gets a user by their ID.
 */
export async function getUserById(userId: number): Promise<TUser | null> {
    const result = await db.query.users.findFirst({
        where: eq(users.id, userId),
    })
    return result ?? null
}

/**
 * Gets a user by their email.
 */
export async function getUserByEmail(email: string): Promise<TUser | null> {
    const result = await db.query.users.findFirst({
        where: eq(users.email, email),
    })
    return result ?? null
}

/**
 * Creates a new user with hashed password.
 */
export async function createUser(
    email: string,
    password: string,
    displayName: string,
): Promise<TUser> {
    const hashedPassword = await hashPassword(password)

    const newUser: TNewUser = {
        email,
        password: hashedPassword,
        display_name: displayName,
        is_onboarded: false,
    }

    const result = await db.insert(users).values(newUser).returning()
    const user = result[0]

    if (!user) {
        throw new Error("Failed to create user")
    }

    return user
}

/**
 * Authenticates a user with email and password.
 * Returns the user if credentials are valid, null otherwise.
 */
export async function authenticateUser(email: string, password: string): Promise<TUser | null> {
    const user = await getUserByEmail(email)
    if (!user) {
        return null
    }

    const isValid = await verifyPassword(password, user.password)
    if (!isValid) {
        return null
    }

    return user
}

/**
 * Marks a user as onboarded.
 */
export async function onboardUser(userId: number): Promise<TUser> {
    const result = await db
        .update(users)
        .set({ is_onboarded: true })
        .where(eq(users.id, userId))
        .returning()

    const user = result[0]

    if (!user) {
        throw new Error("Failed to onboard user")
    }

    return user
}

/**
 * Checks if an email is already registered.
 */
export async function isEmailTaken(email: string): Promise<boolean> {
    const user = await getUserByEmail(email)
    return user !== null
}
